test/common/config/filesystem_subscription_test_harness.h

```diff
@@ -31,7 +34,18 @@ class FilesystemSubscriptionTestHarness : public SubscriptionTestHarness {
   FilesystemSubscriptionTestHarness()
       : path_(TestEnvironment::temporaryPath("eds.json")),
         api_(Api::createApiForTest(stats_store_, simTime())),
-        dispatcher_(api_->allocateDispatcher("test_thread")),
+        dispatcher_([this]() -> Event::DispatcherPtr {
+          auto dispatcher = std::make_unique<Event::MockDispatcher>();
+          EXPECT_CALL(*dispatcher, createFilesystemWatcher_()).WillOnce(InvokeWithoutArgs([this] {
+            Filesystem::MockWatcher* mock_watcher = new Filesystem::MockWatcher();
+            EXPECT_CALL(*mock_watcher, addWatch(path_, Filesystem::Watcher::Events::MovedTo, _))
+                .WillOnce(
+                    Invoke([this](absl::string_view, uint32_t,
+                                  Filesystem::Watcher::OnChangedCb cb) { on_changed_cb_ = cb; }));
```

Comment:
On the surface I'm not super-enthusiastic about this change. It loos a lot more complicated.

I think we should instead build the memory filesystem so we can get the same behavior, and make tests less flaky.

Reply:
We can still inject an in-memory filesystem implementation into these tests, though I would argue that is a bigger refactor outside the scope of this change. I would expect any refactor to eliminate the flakiness in these tests would have to look something similar like what we have here. Other solutions may may have more layered abstractions but ultimately have to do equivalent equivalent steps.

We chose this approach because it injects fake dependencies that make the test single threaded so we can focus only on testing `FilesystemSubscriptionImpl`. Before this, the dispatcher here would spin up a real libevent event loop and watch for events that came from the real filesystem watcher. The flakiness in these tests comes not only from the watcher sometimes being slow to see the filesystem events but also libevent watching the file descriptor it is given to be notified of events.

Since we've faked all the dependencies completely, we also remove any need for cross-thread synchronization and the pretty much the only code being run is what we've faked or from the class under test. Implementing an in-memory filesystem watcher would still require us to fake the dispatcher to return that implementation and also have to trigger the in-memory filesystem watcher to call its file changes callback. While this would succeed in the making the test work with the correct synchronization, it seems a bit more complicated than what we have here.

If you can let us know your ideas about a better picture about how to simplify this, we're open to making the changes!

Reply:
I think this is OK, given it's limited to just the filesystem xDS subscription test, where we don't actually rely on any physical filesystem properties. I would recommend factoring out this dispatcher setup to a standalone function, so that we don't have constructors that look like JavaScript :-P

Reply:
Should be fixed in latest changes